### PR TITLE
fix: query stdout terminal size to see if the output gose to a tty.

### LINKED ISSUE #433: bug: `--color=automatic` dose not work
When piping output of `eza` to another program or simply redirecting stdout to  another file, color controll characters present in the output. The expected behavior is that colorization is disabled when `eza` detect that it is not printing things to pty/tty. 

To reproduce:

```fish
eza --color=automatic --online | fzf
eza --color=automatic --online > eza-out; cat ./eza-out
```

You can see things like `[1;34mvar[0m`.

In contrast, when using coreutils ls, no color should be there.

- The version of eza being used (`eza --version`): `v0.13.1 [+git]`
- The command-line arguments you are using: `eza --color=automatic --oneline`
- Your operating system and hardware platform: `Arch Linux, x86_64`
- This bug is reproduced on a clean container using the `docker.io/ubuntu:latest` image, so it is not my env/config file causing the problem.

### LINKED ISSUE #434: feat: Do not colorize the output when stdout is not connected to a real terminal when `--color=auto` or `--color=automatic`  is set.
coreutils `ls` and [exa](https://github.com/ogham/exa) has a `--color=auto` option that disables colorization when the output is redirected or piped.  
This is very helpful when you want to combine `eza` with other CLI/TUI tools. I am listing a few use cases

- `eza --color=auto --online *.py| fzf` to make a fuzzy search on the `eza` output.
- `eza --color=auto | rg SOMEFANCYREGEX` to do subsequent search in the result.
- `eza --color=auto | nvim -` to edit the output manually. 
- `eza > eza-ouput` to save the output as a text file.

In these cases, you don't want the ANSI controll characters to be included. 

`ls --color=auto` dose exactly the thing we want. Since `eza` is intended to be a replacement of  `ls`, we should support similar functionality.

----

Relavent issue: https://github.com/eza-community/eza/issues/433

### LINKED ISSUE #441: bug: different behaviour from `ls` when stdout is not a tty.
When I am using `eza` with `xargs -I`, I get different output with `ls`.
```fish
❯ ls
1  2  3

❯ /bin/ls | xargs -I {} echo {} {}
1 1
2 2
3 3

❯ /bin/eza | xargs -I {} echo {} {}
1  2  3 1  2  3
```

It seems that when stdout is not a tty, `ls` will separate each file by `'\n'`
```fish
❯ /bin/ls > ../ls_out
❯ /bin/eza > ../eza_out
❯ diff eza_out ls_out
1c1,3
< 1  2  3
---
> 1
> 2
> 3

❯ /bin/ls > /dev/pts/2
❯ 1  2  3
```

env:
- findutils 4.9.0-3
- coreutils 9.3-1
- eza 0.13.1-1

### PR DESCRIPTION
**Pull request by hehelego**
`terminal_size::terminal_size()` is used to determine if stdout is connected to a real tty/pty. In previous version of `terminal-size`, this method only queries the terminal size for `stdout`. However, it is changed in the following commit
https://github.com/eminence/terminal-size/commit/08f0e73926c11adc3105dbf4eb84dd8c9e6c873a Now the method will use `stdout, stdin, stderr`, trying its best to determine the terminal size. This cannot be used to determine if `eza` output is piped or redirected.
We should use `terminal_size_using_fd` on `stdout.as_raw_fd` instead.
This patch makes `eza --color=auto` behaves like `ls --color=auto`.

Resolves #434 Fixes #433

Also Resolves #441 as well.

### CODE REVIEW

**Comment by PThorpe92** [comment]
This looks good, Thanks for the contribution!

I see that this causes some issues on Windows though, take a look at the test results and it shouldn't be too hard to set up a solution. If you don't have access to a windows machine, I can try to grab one from work tomorrow and figure it out.

**Comment by hehelego** [comment]
> Member

Sorry, I don't have a machine with MS Windows installed. With that said, I can look into the document of `terminal-size` and try to fix them.

**Comment by PThorpe92** [comment]
It looks like it even suggests a solution in the error output. No worries, I can get a windows box tomorrow and I'm sure it's just specifying a windows specific method. Shouldn't be too tricky.

**Comment by hehelego** [comment]
CI/CD tasks passed, but I am uncertain about whether this works on real windows machine.

**Review by cafkafk — CHANGES_REQUESTED** [review]
The code looks good to me, and testing seems to show it as working.

I'd prefer if we first got a windows user to test it out as well. Also, from #433 I got the impression that this wasn't documented that well, maybe it would be a good idea to update completions, man page, and readme to reflect this behavior. 

Also, I'd like to pair this with a test case. I'm assuming something like testing `eza --color=automatic --online > eza-out; cat ./eza-out` (as mentioned in #433) and ensuring it doesn't change following this PR would be an option, but I wonder if you had more ideas.

**Comment by hehelego** [comment]
Hi @cafkafk , thanks for pointing out these needy updates. I didn't realize the need for including new testcases and improving documentation.
I'd like to contribute more code/docs and push this forward. I can help revise the man page / shell completion tomorrow.
However, I am unfamiliar with the way you do functionality-level testing (the things in `tests/cmd/`). I'll have to spend some time doing it.

**Comment by cafkafk** [comment]
> Hi @cafkafk , thanks for pointing out these needy updates. I didn't realize the need for including new testcases and improving documentation. I'd like to contribute more code/docs and push this forward. I can help revise the man page / shell completion tomorrow. However, I am unfamiliar with the way you do functionality-level testing (the things in `tests/cmd/`). I'll have to spend some time doing it.

No worries, take as much time as you need! If you have any questions, feel free to ask here, or pop into the matrix chat.

**Review by MartinFillon — APPROVED** [review]

**Comment by daviessm** [comment]
@hehelego I'm just checking this out on Windows. As far as I can tell, redirecting to a file with and without `--color=auto` works as I'd expect. It also worked before though, were we expecting to see any difference on Windows? I may be looking at the wrong thing 😎

**Comment by hehelego** [comment]
> @hehelego I'm just checking this out on Windows. As far as I can tell, redirecting to a file with and without `--color=auto` works as I'd expect. It also worked before though, were we expecting to see any difference on Windows? I may be looking at the wrong thing 😎

Hi @daviessm I am not sure what is the behavior on windows before this patch. But as long as this patch works as expected, I think we are fine.

I spotted the unexpected behavior on Linux platform (see the related issue threads for details). I tried to correct the bug by query `terminal_size_using_fd` of `stdout.as_raw_fd`, however this dose not work for windows (they use handles instead of file descriptors). Later on, I tried to use `terminal_size_using_handle` on `STD_OUTPUT_HANDLE`, expecting that to work on windows.   
We were waiting to test if that actually works on real Windows installation.

**Review by daviessm — APPROVED** [review]
Yeah, looks like the original code always worked for Windows and this doesn't break it so LGTM for Windows 👍

**Comment by hehelego** [comment]
BTW, I fish shell has a special builtin function `fish_update_completions` which extracts completions from manpages.  
If we are keeping the manpage `exa(1)` up to date, there is no need to keep a separated and synchronized fish completion

**Comment by hehelego** [comment]
> Yeah, looks like the original code always worked for Windows and this doesn't break it so LGTM for Windows 👍

Great. I'll keep working on manpages improvement and test case introduction.

**Comment by daviessm** [comment]
> BTW, I fish shell has a special builtin function `fish_update_completions` which extracts completions from manpages. If we are keeping the manpage `exa(1)` up to date, there is no need to keep a separated and synchronized fish completion

Yep, once the `clap` PR #197 gets merged we can autogenerate completions for multiple shells from that - see https://github.com/eza-community/eza/issues/400#issuecomment-1731026101

**Comment by hehelego** [comment]
I'll need some help. Looks like `trycmd` dose not support complicated CLI testing involving shell, redirection and pipeline.
We may need something like `snapbox` for that.

**Comment by PThorpe92** [comment]
Interesting, I've never heard of it. I'm going to turn this into a discussion post, and look into snapbox.

Feel free to come by the matrix channel if you want, we can see about how to proceed here particularly

**Comment by hehelego** [comment]
Wait, I think I've mis-read that issue. They actually build another testing framework based on `snapbox`.
I've just looked into the document of `snapbox` and seems no easy way to test IO redirecting and pipelining.. 
Sorry for that.

**Comment by hehelego** [comment]
I think we can move forward to merge this PR first. Finding a proper way to test eza's behavior when redirection and pipelining are involved should be a separate task.

**Comment by cfxegbert** [comment]
Why not use the trait [IsTerminal](https://doc.rust-lang.org/stable/std/io/trait.IsTerminal.html#)?

**Comment by hehelego** [comment]
> Why not use the trait [IsTerminal](https://doc.rust-lang.org/stable/std/io/trait.IsTerminal.html#)?

Thanks for pointing it our, exactly the thing we need.

**Comment by hehelego** [comment]
> I think we can move forward to merge this PR first. Finding a proper way to test eza's behavior when redirection and pipelining are involved should be a separate task.

Update: I've come up with a plan to test this using `snapbox`. I'll try to implement it soon.
We can merge this PR after the test are available.

**Comment by daviessm** [comment]
I'll give the new code a whiz on Windows later today.

**Review by daviessm — APPROVED** [review]
Windows still looks good

**Comment by hehelego** [comment]
Yep. I fiind another API misuse: https://github.com/eza-community/eza/blob/afe832b943d328a20f43c6fa69dd8588cf3285ef/src/output/mod.rs#L56

`terminal_size::terminal_size()` will query `stderr` and `stdin` when `stdout` is not connected to a tty/pty.

**Comment by hehelego** [comment]
I cannot test this with `std::process:Command` or `snapbox::cmd::Command`. They use pipe to capture stdout of the spawned process. I have no way to get the "normal" output (when you directly run `eza` from a shell session).

Does anyone have any idea on how should we do this?

**Comment by hehelego** [comment]
I cannot test this with `std::process:Command` or `snapbox::cmd::Command`. They use pipe to capture stdout of the spawned process. I have no way to get the "normal" output (when you directly run `eza` from a shell session).

Does anyone have any idea on how should we do this?

**Comment by hehelego** [comment]
Alright, this thread is becoming too long to read. I shall give a quick summary of the current progress:

- The root cause is identified: An API misuse. `terminal_size::terminal_size()` queries the terminal size of `stdout, stdin, stderr` while we only want size of the terminal connected to `stdout`.
- We've fixed

**Review by Ultra980 — APPROVED** [review]
LGTM!

File: README.md
Comment (MartinFillon) [inline README.md:337]:
```suggestion
- Valid **--co{u}lor** options are **always**, **automatic** (or **auto** for short), and **never**.
```

**Review by MartinFillon — CHANGES_REQUESTED** [review]

File: README.md
Comment (hehelego) [reply README.md:337]: Sorry for not followinig what are already presented in `README.md`.  
We have the following lines writing `colo[u]r`. Definitely should conform to this.

```markdown
- **-F**, **--classify**: display type indicator by file names
- **--colo[u]r**: when to use terminal colours
```

Comment (MartinFillon) [reply README.md]: it is okay everyone make mistakes

**Review by cafkafk — CHANGES_REQUESTED** [review]
I can't push to your branch for some reason, but I rebased and added a breaking notifiers (`!` and `BREAKING CHANGE:`) to the first commit.

```
commit 9a3b1cbd66a07ae0208b99418db0bb2598303582
Author: hehelego <hahalegao@gmail.com>
Date:   Wed Sep 27 18:05:39 2023 -0500

    fix!: query stdout terminal size to see if the output goes to a tty.
    
    `terminal_size::terminal_size()` is used to determine if stdout is
    connected to a real tty/pty. In previous version of `terminal-size`,
    this method only queries the terminal size for `stdout`.
    However, it is changed in the following commit
    https://github.com/eminence/terminal-size/commit/08f0e73926c11adc3105dbf4eb84dd8c9e6c873a
    Now the method will use `stdout, stdin, stderr`, trying its best to
    determine the terminal size. This cannot be used to determine if `eza`
    output is piped or redirected.
    We should use `terminal_size_using_fd` on `stdout.as_raw_fd` instead.
    
    Resolves #434 Fixes #433
    
    BREAKING CHANGE: This is a change in behaviour from before, and so we
    have marked it as breaking. In practice, it probably won't break
    anything except if you're doing something really weird (more power to you!).
```

**Review by cafkafk — APPROVED** [review]
I'm just gonna fix this after merge, still needs a follow up with tests

**Pull request merged**
